### PR TITLE
[DUOS-226][risk=low] Migrate DARs from mongo to postgres

### PR DESCRIPTION
**Pull request by rushtong**
## Addresses
https://broadinstitute.atlassian.net/browse/DUOS-226

See https://broadinstitute.atlassian.net/browse/DUOS-604 for cleanup work

## Changes
* One-off groovy scripts to execute migration
* Copy all DARs from mongo to postgres
* Temporary admin endpoints to facilitate migration
* Read all DAR info from postgres now, aside from migration apis
* Use reference id instead of id as much as possible. Remaining usages will be dealt with when we address the partial dar collection.
* Re-implement some native mongo things such as create date, sorting, date conversions, etc. Sprinkled throughout the codebase
* More tests for service and dao calls.

## Tests
The following workflows are manually verified.

* Export dars from mongo -> import to postgres using groovy script
* Researcher
  * View dar console 
  * Create new dar 
  * Submit new dar 
  * Create partial dar and save 
  * Submit partial dar 
  * Review dar 
  * Cancel dar 
* Admin
  * View dar manage page 
  * Create new dar election 
  * Create migrated dar election 
  * Cancel dar election 
  * Create canceled dar election 
  * View dar summary 
  * View dar election 
  * After chair final vote, election shows in reviewed cases 
* Admin - Statistics
  * DUL: All Cases 
  * DUL: Reviewed Cases 
  * DAR: Should data be granted to this applicant?
    * All Cases 
    * Reviewed Cases 
  * DAR: Was the research purpose accurately converted to a structured format? 
    * All Cases 
    * Reviewed Cases 
  * DAR: Evaluation of automated matching in comparison with the Data Access Committee decision 
    * Cases reviewed by automated matching  
    * Agreement between automated matching and Data Access Committee 
* Member
  * View dar page (filtered by DAC) 
  * View dar after new election created 
  * Vote on migrated dar 
  * Vote on new dar 
  * DAR page shows correct votes 
  * Edit vote 
* Chair
  * View dar page (filtered by DAC) 
  * Vote as member  
  * Vote as chair 
  * Collect votes on dar 
  * Final vote on dar

### CODE REVIEW

File: src/main/java/org/broadinstitute/consent/http/service/DataAccessRequestService.java [hunk 1]
```diff
@@ -143,14 +146,121 @@ public Integer getTotalUnReviewedDars(AuthUser authUser) {
...
+
     private Map<String, Election> createAccessRequestElectionMap(List<Election> elections) {
         Map<String, Election> electionMap = new HashMap<>();
         elections.forEach(election -> electionMap.put(election.getReferenceId(), election));
         return electionMap;
     }
 
+    /**
+     * Convenience method during transition away from `Document` and to `DataAccessRequest`
+     */
```
Comment (rushtong) [inline src/main/java/org/broadinstitute/consent/http/service/DataAccessRequestService.java:169]: Add todos here so we know which ones need to be deleted in subsequent cleanup PRs.

File: src/main/java/org/broadinstitute/consent/http/service/DataAccessRequestService.java [hunk 2]
```diff
@@ -209,34 +319,39 @@ public Integer getTotalUnReviewedDars(AuthUser authUser) {
             }
             requestsManage.add(darManage);
         });
-
         return populateElectionInformation(
-                populateDacInformation(requestsManage),
-                referenceIdElectionMap,
-                user);
+                populateDacInformation(requestsManage), referenceIdElectionMap, user);
```
Comment (rushtong) [inline src/main/java/org/broadinstitute/consent/http/service/DataAccessRequestService.java]: Revert this change

File: src/main/java/org/broadinstitute/consent/http/service/DataAccessRequestService.java [hunk 3]
```diff
@@ -209,34 +319,39 @@ public Integer getTotalUnReviewedDars(AuthUser authUser) {
...
     }
 
     /**
      * Return a cloned, immutable list of DataAccessRequestManage objects with election and vote information populated
      */
     private List<DataAccessRequestManage> populateElectionInformation(List<DataAccessRequestManage> darManages, Map<String, Election> referenceIdElectionMap, DACUser user) {
         Collection<Election> elections = referenceIdElectionMap.values();
-        List<Integer> electionIds = referenceIdElectionMap.values().stream().
+        List<Integer> electionIds = referenceIdElectionMap.values().
+                stream().
```
Comment (rushtong) [inline src/main/java/org/broadinstitute/consent/http/service/DataAccessRequestService.java]: revert this

File: src/main/java/org/broadinstitute/consent/http/service/DataAccessRequestService.java [hunk 4]
```diff
@@ -335,21 +449,18 @@ private String consolidateDataSetElectionsResult(List<Election> datasetElections
     }
 
     private List<Document> getUnReviewedDarsForUser(AuthUser authUser) {
+        List<Document> activeDars = getAllDataAccessRequestsAsDocuments().stream().
+                filter(d -> !ElectionStatus.CANCELED.getValue().equalsIgnoreCase(d.getString(DarConstants.STATUS))).
+                collect(Collectors.toList());
```
Comment (rushtong) [inline src/main/java/org/broadinstitute/consent/http/service/DataAccessRequestService.java:482]: Note: Add a story to allow for querying into the json in postgres so we don't have to stream over all DARs.

Comment (rushtong) [reply src/main/java/org/broadinstitute/consent/http/service/DataAccessRequestService.java:482]: Added https://broadinstitute.atlassian.net/browse/DUOS-609

File: src/main/java/org/broadinstitute/consent/http/db/DataAccessRequestDAO.java [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+import org.jdbi.v3.sqlobject.statement.SqlQuery;
+import org.jdbi.v3.sqlobject.statement.SqlUpdate;
+import org.jdbi.v3.sqlobject.transaction.Transactional;
+
+import java.util.List;
+
+@RegisterRowMapper(DataAccessRequestMapper.class)
+public interface DataAccessRequestDAO extends Transactional<DataAccessRequestDAO> {
+
+    @SqlQuery("select * from data_access_request")
```
Comment (mmorgantaylor) [inline src/main/java/org/broadinstitute/consent/http/db/DataAccessRequestDAO.java]: nit: use standard SQL capitalization, e.g. `SELECT * FROM data_access_request` - would help with readability

Comment (rushtong) [reply src/main/java/org/broadinstitute/consent/http/db/DataAccessRequestDAO.java]: Will do, thank you!

**Comment by rushtong** [comment]
Merging this - we need it for the next release.
